core/headerchain.go

```diff
@@ -105,302 +122,212 @@ func NewHeaderChain(chainDb ethdb.Database, config *params.ChainConfig, engine c
-		lastHash   = headers[0].ParentHash[types.QuaiNetworkContext]          // Last imported header hash
-
-		newTd = ptd // Total difficulty of inserted chain
-
-		lastHeader    *types.Header
-		inserted      []numberHash // Ephemeral lookup of number/hash for the chain
-		firstInserted = -1         // Index of the first non-ignored header
-	)
-
-	batch := hc.chainDb.NewBatch()
-	parentKnown := true // Set to true to force hc.HasHeader check the first iteration
-	for i, header := range headers {
-		var hash common.Hash
-		// The headers have already been validated at this point, so we already
-		// know that it's a contiguous chain, where
-		// headers[i].Hash() == headers[i+1].ParentHash
-		if i < len(headers)-1 {
-			hash = headers[i+1].ParentHash[types.QuaiNetworkContext]
-		} else {
-			hash = header.Hash()
+		// Trim the branch before dequeueing
+		commonHeader := hc.findCommonHeader(hc.heads[0])
+		err = hc.trim(commonHeader, hc.heads[0])
+		if err != nil {
+			return err
 		}
 
-		newTd, err = hc.CalcTd(header)
-		if err != nil {
-			return &headerWriteResult{}, errors.New("error calculating the total td for the header")
+		// dequeue
+		hc.heads[0] = nilHeader
+		hc.heads = hc.heads[1:]
+	}
+	// Add to the heads queue
+	hc.heads = append(hc.heads, block.Header())
+
+	// Sort the heads by number
+	sort.Slice(hc.heads, func(i, j int) bool {
+		return hc.heads[i].Number[types.QuaiNetworkContext].Uint64() < hc.heads[j].Number[types.QuaiNetworkContext].Uint64()
+	})
+
+	return nil
+}
+
+// SetCurrentHeader sets the in-memory head header marker of the canonical chan
+// as the given header.
+func (hc *HeaderChain) SetCurrentHeader(head *types.Header) error {
+	hc.currentHeader.Store(head)
+	hc.currentHeaderHash = head.Hash()
+	headHeaderGauge.Update(head.Number[types.QuaiNetworkContext].Int64())
+
+	//Update canonical state db
+	//Find a common header
+	commonHeader := hc.findCommonHeader(head)
+	parent := head
+
+	// Delete each header and rollback state processor until common header
+	// Accumulate the hash slice stack
+	var hashStack []*types.Header
+	for {
+		if parent.Hash() == commonHeader.Hash() {
+			break
 		}
 
-		number := header.Number[types.QuaiNetworkContext].Uint64()
-
-		// If the parent was not present, store it
-		// If the header is already known, skip it, otherwise store
-		alreadyKnown := parentKnown && hc.HasHeader(hash, number)
-		if !alreadyKnown {
-			// Irrelevant of the canonical status, write the TD and header to the database.
-			rawdb.WriteTd(batch, hash, number, newTd)
-			hc.tdCache.Add(hash, newTd)
-
-			rawdb.WriteHeader(batch, header)
-			inserted = append(inserted, numberHash{number, hash})
-			hc.headerCache.Add(hash, header)
-			hc.numberCache.Add(hash, number)
-			if firstInserted < 0 {
-				firstInserted = i
-			}
+		// Delete the header and the block
+		rawdb.DeleteCanonicalHash(hc.headerDb, parent.Number64())
```

Comment:
```// Delete the header and the block```
Is this comment wrong? I think you only remove the header from the canonical DB. No need to delete the header or block from the header/block DBs